### ISSUE TITLE
Garbage Collector for immutable `ConfigMap`s/`Secret`s which are no longer in-use 

### ISSUE THREAD

**Issue by rfranzke** [gardener/gardener-resource-manager#104]
**How to categorize this issue?**

/area robustness quality
/kind enhancement
/priority normal

**What would you like to be added**:
GRM could add a hash suffix to the names of `ConfigMap`s or `Secret`s that are mounted by `Deployment`s/`DaemonSet`s/etc. (it would have to adapt the names also in the `.volumes[]` section). The hash could be a checksum of the `ConfigMap`'s/s`Secret`'s `.data` map.

This should be done by default, however, it should be possible to
- disable this entire behaviour via a CLI parameter
- opt-out of this behaviour for particular resources via an annotation
- opt-out of this behaviour for all resources of a `ManagedResource` via a field in the `.spec`

Additionally, the `ConfigMap`s/`Secret`s should be marked [immutable](https://kubernetes.io/docs/concepts/configuration/configmap/#configmap-immutable).

**Why is this needed**:
GRM creates the Kubernetes resources managed by a `ManagedResource` in parallel. This can lead to issues when e.g. a `Deployment` is mounting a `ConfigMap`/`Secret` and the `Deployment` is updated before the `ConfigMap`/`Secret` and/or the update of the `ConfigMap`/`Secret` fails for whatever reason. In rare cases it can happen that a new pod is already rolled out and started but still mounting the old `ConfigMap`/`Secret` data. Eventually, GRM will also succeed updating the `ConfigMap`/`Secret`, however, the pod is already running and has started with the old configuration data. This can lead to issues when the new pod is incompatible with the old configuration data.

Another problematic case is when the kubelet's `ConfigMap`/`Secret` cache is stale. Also, very rarely, we had observed this in the past, i.e., the kubelet starts a pod but mounts the old/outdated configuration data. Eventually, the kubelet's cache will be up-to-date again, but the pod might have already been started and read the old config.

With above implementation (hash suffix) both problematic scenarios cannot happen anymore because a new pod would only start with the new `ConfigMap`/`Secret`.

**Implementation considerations**:
We would need a mechanism to cleanup the old/no longer used `ConfigMap`s/`Secret`s, perhaps with a separate controller.

/cc @timebertt @amshuman-kr @shreyas-s-rao @mvladev

**Comment by amshuman-kr** [comment]
Thanks for the proposal @rfranzke, @timebertt. Seems sound. What would be the impact on the `ManagedResource` status which lists the deployed resources? Will it list the configmap and secrets with or without the hash suffix?

**Comment by rfranzke** [comment]
It should list them only WITH the hash suffix because this is what is actually deployed.

**Comment by rfranzke** [comment]
After a "small PoC" end of last week and a thorough discussion with @timebertt yesterday we decided to not implemented it in the way proposed above. We figured that there probably isn't an elegant generic solution to this problem (from GRM's point of view) for the following reasons:

1. After GRM has computed a checksum suffix for `ConfigMap` or `Secret` names it can only replace "well-known" occurrences of the respective references (e.g., in `core/v1.PodTemplate.spec.volumes` or `core/v1.PodTemplate.spec.containers[].env`). While this might work in some (maybe even many?) cases it will fail for e.g. deployments which don't mount the `ConfigMap`/`Secret` but refer to it via a CLI parameter (like the [nginx-ingress-controller](https://github.com/gardener/gardener/blob/master/charts/seed-bootstrap/charts/nginx-ingress/templates/controller-deployment.yaml#L60)). It will also fail for custom resources whose fields are "even more unknown" to us.
2. This means that such deployments would need an opt-out option of this GRM behaviour and (re-)implement the same behaviour (checksum suffix) on their own.
3. We think that this might increase complexity and makes developing such components harder ,especially for new or non-frequent contributors.

Instead, we decided for an alternative approach which is adapting all our components to no longer use a checksum annotation in their `core/v1.PodTemplate` but rather a checksum suffix for the names of their `ConfigMap`s/`Secret`s. Additionally, we set `.immutable=true`. This may sound "scary"/like a lot of effort, but actually it isn't. We already have these "specialities" today (checksum annotation), so we are already "touching" each and every component/chart. Now, we are just changing the mechanism, so it's not getting worse.

Still, there are two more considerations/requirements arising out of this:

1. Today, when the checksum suffix would change, GRM would immediately delete the "old" `ConfigMap`/`Secret`. This will prevent existing pods from starting up again (e.g., after a node got evicted or a deployment was scaled) because the resources are already deleted from the system. This can lead to downtimes if the roll-out of the new pods is not progresssing or delayed for whatever reason. Consequently, GRM needs to become more intelligent when it comes to deleting "old resources": GRM would have a new, separate controller watching `Pod`s and their `.spec.volumes` and `.spec.containers[].env`. If it detects a `ConfigMap`/`Secret` reference and this `ConfigMap`/`Secret` is marked as `immutable` then it does not delete it immediately. Instead, it waits until no such pod exists anymore before it deletes those resources. Obviously, this is just an "approximation" because we still cannot handle references in fields other than those "well-known" to us, however, we considered it good enough for now and the use-cases we have.
2. Gardener uses the GRM only for deploying resources into the shoot cluster (system components, optional addons). Resources in the seed (control plane, monitoring/logging stack, ...) are still deployed "manually"/directly via the Kubernetes client. Consequently, if we would implement the above (checksum suffix) for the control plane components then we would also need a separate cleanup logic for the old/stale `ConfigMap`/`Secret`s. Probably, now the time has come to change how resources in the seed are deployed and also leverage GRM for it. This saves us from the mentioned "code duplication".

/title Old ConfigMaps/Secrets should only be deleted if no longer "in-use"
/assign

**Comment by timebertt** [comment]
> 1. GRM would have a new, separate controller watching `Pod`s

We should use metadata-only watches for this, as we will otherwise increase the cache size in grm for shoots having lots of Pods. Ref https://github.com/kubernetes-sigs/controller-runtime/pull/1174

**Comment by rfranzke** [comment]
Thanks @timebertt. Without being the details, just to clarify on this: The reconciliation loop of this new controller would need to read the `Pod`'s specification in order to evaluate whether it has to do something or not. Would this result in a "real" call because the cache does only contain the metadata?

**Comment by timebertt** [comment]
Yes, the watch only transports metadata (basically notifications about changes, without the details). The controller would then need to get the Pod using a dedicated call.
Yes, the advantages/disadvantages (i/o vs mem) need to be weighted here. But I'm just a bit concerned that such a Pod watch/cache will open the door for shoot owners to blow up grm's caches again (similar to https://github.com/gardener/gardener-resource-manager/issues/92).

**Comment by voelzmo** [comment]
I didn't find a reference to it in this issue, so I'm leaving this here: https://github.com/kubernetes/enhancements/pull/948
You most likely are aware of the discussions in there regarding immutability and deletion/garbage collection of Secrets/ConfigMaps and the timing issues around "depending on what is deployed first you'll get different results", so I'm cross-linking it for reference.

**Comment by rfranzke** [comment]
While starting to sketch out a first implementation of https://github.com/gardener/gardener-resource-manager/issues/104#issuecomment-771388854, @timebertt and myself got some concerns with the approach.

First of all, we are still not too happy with the “approximation” mentioned in above comment, i.e., a general “garbage collector” cannot find out whether a `ConfigMap` or `Secret` is still in use by a `Pod` or something else (e.g., some pods like `nginx-ingress` only refer to their `ConfigMap` via CLI flags in the command of a container; in other cases `ConfigMap`s/`Secret`s might be connected with custom resources whose specs are unknown). While such corner cases might be negligible today, it is already technical debt with the potential to break us in the future and causing follow-up efforts.

Secondly, we don’t leverage `gardener-resource-manager` for many things today (e.g., the shoot control plane components are deployed by gardenlet and not actively reconciled). In above comment, we describe that we should switch this to using GRM also for those components (https://github.com/gardener/gardener/issues/3485). However, we don’t want to go ahead with it yet because (a) this would result in a lot of `ManagedResource`s in the seed cluster and we don’t know the performance implications of such huge increase, and (b) this would mean that we duplicate a lot of data in `Secret`s (e.g., `ConfigMap`s containing Grafana dashboard would be part of a `Secret` referenced in a `ManagedResource` before actually be created as `ConfigMap`s in the seed).

The first “quick & dirty” sketch (https://github.com/gardener/gardener-resource-manager/compare/master...rfranzke:enh/immutable) still follows above approach in the following way:

1. gardenlet would compute unique names for the relevant `ConfigMap`s/`Secret`s and label them with `resources.gardener.cloud/consider-garbage-collection=true` (in both seed and shoot clusters).
2. A new garbage-collector controller in the `gardener-resource-manager` (GRM) would regularly list all such `ConfigMap`s/`Secret`s and try to find out whether they are still in use by
    1. listing all `Deployment`s, `ReplicaSet`s, `StatefulSet`s, `Job`s, `Pod`s.
    2. checking their pod template specs for referenced `ConfigMap`s/`Secret`s in their `volumes` or `envFrom` fields.
3. All `ConfigMap`s/`Secret`s with above label which weren’t found in any pod template specs would be deleted.

_Please note that there is a central GRM per seed and a dedicated GRM per shoot, hence, having such garbage-collector controller would work, though, it doesn’t fit too nicely._

The outlined solution results in quite a lot of LIST calls against the API servers. While we might be able to lengthen the period, LISTing is always expensive and should be avoided if possible, especially in large clusters like those we run.
Also, it still has the downside that it would still just be an “approximation” as explained above.

Instead, @timebertt and myself envision an alternative approach:

1. gardenlet labels pod templates with `gardener.cloud/consider-for-volume-consistency-check=true` (or something similar) for components mounting `ConfigMap`s/`Secret`s relevant for such checks. Also, it annotates the pod templates with the expected checksums for the `ConfigMap`s/`Secret`s (already done today).
2. The `gardener-seed-admission-controller` (GSAC) (already running centrally per seed) serves a webhook endpoint for `CREATE pod` events. The webhook will be registered in the seed and in all shoot clusters.
3. When a pod with above mentioned label is created, GSAC injects an init container which mounts all `ConfigMap`/`Secret` volumes. It computes the checksum of the mounted data and compares it against the expected checksums from the pod template annotations. Only when they are equal it exits successfully, allowing the actual workload containers to start up.

This approach has the advantage that it works for all cases as the client (gardenlet) controls when to do what (i.e., for which components which volumes need to be considered for consistency checks). Additionally, we don’t need to compute unique names for `ConfigMap`s/`Secret`s, hence, we don’t need any garbage collector, hence, we can save quite some WATCH and LIST calls.
Also, the “infrastructure” (coding-wise) already exists, i.e., GSAC already runs in HA-mode in all seeds.

**Comment by amshuman-kr** [comment]
Very interesting approach!

I have one question below.

> When a pod with above mentioned label is created, GSAC injects an init container which mounts all ConfigMap/Secret volumes. It computes the checksum of the mounted data and compares it against the expected checksums from the pod template annotations. Only when they are equal it exits successfully, allowing the actual workload containers to start up.

If the deployment was rolled first and the configmap/secret updated later, at first the init container would fail but would the init-container or the pod be restarted automatically for it to eventually succeed when the configmap/secret is updated?

**Comment by rfranzke** [comment]
> If the deployment was rolled first and the configmap/secret updated later, at first the init container would fail but would the init-container or the pod be restarted automatically for it to eventually succeed when the configmap/secret is updated?

Yes, at least this is the assumption. We need to try this out, I guess there are two options: Either the init container waits for some time (to allow kubelet to mount the updated configmap/secret) and then tries to compare the checksums again (in this case, it would be running fine all the time), or the init container fails immediately when the checksum comparison fails (in this case, kubelet would restart it automatically). If there are already preferences let me know :)

**Comment by amshuman-kr** [comment]
> I guess there are two options: Either the init container waits for some time (to allow kubelet to mount the updated configmap/secret) and then tries to compare the checksums again (in this case, it would be running fine all the time), or the init container fails immediately when the checksum comparison fails (in this case, kubelet would restart it automatically). If there are already preferences let me know :)

No preferances as such. Just remembered some bad experience with init-containers and container `restartPolicy` from years ago when we were trying out approaches for etcd data validation in the backup sidecar. I remember that failing init-containers didn't work well with the restart policy. But that might been a different case.

Also, one more question, would this approach also work for configmap/secret being used as environment variables?

**Comment by rfranzke** [comment]
> Also, one more question, would this approach also work for configmap/secret being used as environment variables?

Good question! Yes, in this case, the init container would still mount the configmaps/secrets appearing in other containers' env variables and check for the checksums :)

**Comment by vlerenc** [comment]
@rfranzke I was confused how the third approach (https://github.com/gardener/gardener-resource-manager/issues/104#issuecomment-874679849) would resolve the issue with not well-known references, such as [nginx-ingress-controller](https://github.com/gardener/gardener/blob/master/charts/seed-bootstrap/charts/nginx-ingress/templates/controller-deployment.yaml#L60) (btw, also the second approach (https://github.com/gardener/gardener-resource-manager/issues/104#issuecomment-771388854) seemed not to "fix" it even though that was the argument against the first approach in the initial description). But now you wrote:

> [I]n this case, the init container would still mount the configmaps/secrets appearing in other containers' env variables and check for the checksums :)

How does/should that work?

Frankly, I don't see how the problem with not  well-known references can be fixed by any other than an explicit approach that ties together what belongs together. What am I not understanding here?

**Comment by rfranzke** [comment]
> btw, also the second approach (#104 (comment)) seemed not to "fix" it even though that was the argument against the first approach in the initial description

In the first approach (initial issue description), we wanted GRM to do all of this automagically (computing unique names, replacing the original name for all occurrences, etc.), without the client (gardenlet) to care. The argument against it was that this won't work in all cases. Also, another argument against it was that we don't use GRM for everything yet (and maybe don't want?), so we can't achieve it easily anyways (at least today).

In the second approach, the client (gardenlet) has to take care (computing unique names, replace the well-known occurrences with the new names), though, it still has the same problems (a generic garbage collector cannot know whether a `ConfigMap`/`Secret` is still in use in all cases). Additionally, this garbage collector could potentially produce unnecessary load on the API servers. Those two downsides were motivating us to think about yet another alternative.

The third approach (init container via webhook in GSAC) was the result of this brain-storming. You're right, cases like the `nginx-ingress-controller` are not "fixed" just by going with this approach, it must be explicit. Still, we prefer this approach over the first two as it's much less complex (no unique secret names with hash suffixes), easier to implement (small binary computing and comparing checksums plus a small webhook injecting the init container), easier to operate (it's always clear which `ConfigMap`/`Secret` is mounted/no potential for confusion), and less load/API calls (nothing to clean up).

**Comment by vlerenc** [comment]
OK @rfranzke, so it doesn't resolve that particular problem. Yes, I understood, some of the other problems are resolved/mitigated, but all of that is not ideal.

When you write "must be explicit" then you mean what exactly?

**Comment by rfranzke** [comment]
> but all of that is not ideal.

Why not, what are your concerns?

> When you write "must be explicit" then you mean what exactly?

I need to get my hands dirty with it, but a first rough idea could be as follows:

1. For the "standard" cases (`ConfigMap`/`Secret` is mounted via `volume` or referred via `envFrom`), the gardenlet would (just like today) annotate the respective pod template specs with the expected checksum. In addition to today, it would add a label to the pod template spec (`consider-for-data-integrity-check=true`) so that the webhook in GSAC gets triggered.
2. For the "special" cases (`ConfigMap`/`Secret` is not mounted but referenced via CLI flags (e.g., in the `nginx-ingress-controller` case)), the gardenlet would do the same like in 1. Additionally, it adds another annotation so that the webhook in GSAC knows which `ConfigMap`/`Secret` is needs to mount to the init container, e.g. `data-integrity-check/secret:my-secret`.

That's it. Most of it is already implemented (checksum annotations), hence, we just need to add the label to the templates and consider special cases like `nginx-ingress` (the only one coming to my mind today, but will check).

With the second approach (unique names + generic garbage-collector), despite from the doubts we already have, I don't even have an idea how it could handle special cases like `nginx-ingress` in order to find out whether a `ConfigMap`/`Secret` is still in use, even if the gardenlet does something explicitly.

**Comment by vlerenc** [comment]
In addition @rfranzke, the failing init container, isn't that potentially a threat, too? I mean, it prevents the pod to start, so that sounds unsafe.

Overall, it "feels" not right, doesn't resolve the very same issues that were used to discredit the previous approaches, creates potential hazards.

**Comment by rfranzke** [comment]
> In addition @rfranzke, the failing init container, isn't that potentially a threat, too? I mean, it prevents the pod to start, so that sounds unsafe.

Better it doesn't start than it starts with stale data, or? :-) Also, when we update pods, we usually use the 'rolling update' mode, i.e., the old pod will still run until the new one is ready.

> Overall, it "feels" not right, doesn't resolve the very same issues that were used to discredit the previous approaches, creates potential hazards.

Hm, it does resolve all our issues and achieves the initial goal with minimal adaptations required in gardenlet, right? We'll get data integrity checks ensuring that pods only start with the version of the data in a `ConfigMap`/`Secret` that we want them to have. Do I miss something?

**Comment by amshuman-kr** [comment]
> the init container would still mount the configmaps/secrets appearing in other containers' env variables and check for the checksums :)

If I got it right, the webhook would mutate the `volumes` in the pod template to add any configmaps/secrets that were solely used as environment variable, right?

**Comment by rfranzke** [comment]
Yes, exactly @amshuman-kr!

**Comment by vpnachev** [comment]
The webhook injecting the init container have to have in mind that other mutating webhooks can (un)mount secrets/configmaps to the pod. Therefore, it should be invoked last, if possible. See https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#reinvocation-policy for more details.

**Comment by rfranzke** [comment]
After another sync with @vlerenc and @timebertt we were still unhappy with the init container approach:

* A new container cannot start if the mounted `ConfigMap`/`Secret` is outdated (that's fine and exactly what we want).
* However, rolling updates might be slowly progressing/blocked for whatever reason (e.g., new pod does not start-up). In such case, the `ConfigMap`/`Secret` was already updated. If the old, still running pod is now deleted (e.g., due to VPA scaling) then the newly injected init container will prevent the container from starting up again.
* Even worse, if the old, still running pod is not deleted but a container only fails then it gets restarted by kubelet without a new execution if the init container. Then the old container will use the new `ConfigMap`/`Secret` which is might be problematic.

We discussed again the garbage collector approach and having unique names for `ConfigMap`s/`Secret`s (plus making them immutable) and decided to go for it as follows:

GRM features a GC controller which
1. lists `ConfigMap`s/`Secret`s with some well-known label in all namespaces.
2. lists all `Deployment`s, `StatefulSet`s, etc. (all workload resources)
3. iterates over the annotations of such workload resources
4. If the annotation key follows the scheme `reference/{configmap,secret}_<name>` then GRM considers the respective `ConfigMap`/`Secret` as "in-use" (because a workload resources was found).
5. All remaining `ConfigMap`s/`Secret`s are considered as "not in-use" and deleted.

Please note that this also works for cases like nginx-ingress as discussed [above](https://github.com/gardener/gardener-resource-manager/issues/104#issuecomment-875617147).

It's an opt-in approach and the client must pro-actively

1. add the well-known label to garbage-collectable `ConfigMap`s/`Secret`s
2. use unique names for such `ConfigMap`s/`Secret`s (e.g., based on the hash of the `.data` map)
2. follow the `reference/{configmap,secret}_<name>` scheme and have respective annotation keys in the workload resources using `ConfigMap`s or `Secret`s

In https://github.com/gardener/gardener-resource-manager/issues/104#issuecomment-875606267, we argued

1. that the `LIST` calls might stress the kube-apiserver, however, we can do metadata-only `LIST`s and have a high sync period.
2. > it's much less complex (no unique secret names with hash suffixes),

    This is still true, having unique names with suffixes is more complex.

3. > easier to implement (small binary computing and comparing checksums plus a small webhook injecting the init container)

    With above approach, the implementation is similarly easy with the GC controller.

4. > easier to operate (it's always clear which ConfigMap/Secret is mounted/no potential for confusion)

    True, see above, it gets a bit more complex now during operations.

5. > and less load/API calls (nothing to clean up).

    See above (addressed with metadata-only list calls and high sync periods).

**Comment by amshuman-kr** [comment]
> If the annotation key follows the scheme reference/{configmap,secret}_<name> then GRM considers the respective ConfigMap/Secret as "in-use" (because a workload resources was found).

The `reference/XXX` annotation is expected to be added to the pod templates, right?

**Comment by rfranzke** [comment]
@amshuman-kr Good question, I have clarified this in the docs (section about "what clients need to do"): https://github.com/gardener/gardener-resource-manager/blob/54df0477f804f1655bc233bbbb10747ff52e5932/docs/concepts/garbagecollector.md#how-does-the-garbage-collector-work
